staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller_test.go

```diff
@@ -50,124 +56,120 @@ func TestProcessEncryptionConfig(t *testing.T) {
 			)
 
 			_, _, err := d.processEncryptionConfig(ctx)
-			if testCase.expectError && err == nil {
-				t.Fatalf("expected error but got none")
+
+			gotError := ""
+			if err != nil {
+				gotError = err.Error()
 			}
-			if !testCase.expectError && err != nil {
-				t.Fatalf("expected no error but got %v", err)
+
+			if gotError != testCase.expectError {
+				t.Fatalf("expected error %q but got %q", testCase.expectError, gotError)
 			}
 		})
 	}
 }
 
-func TestWatchEncryptionConfigFile(t *testing.T) {
-	testCases := []struct {
-		name          string
-		generateEvent func(filePath string, cancel context.CancelFunc)
-		expectError   bool
+func TestRun(t *testing.T) {
+	tests := []struct {
+		name     string
+		filePath string
 	}{
 		{
-			name:        "file not renamed or removed",
-			expectError: false,
-			generateEvent: func(filePath string, cancel context.CancelFunc) {
-				os.Chtimes(filePath, time.Now(), time.Now())
-
-				// wait for the event to be handled
-				time.Sleep(1 * time.Second)
-				cancel()
-				os.Remove(filePath)
-			},
+			name:     "with empty config",
+			filePath: "testdata/empty_config.yaml",
 		},
 		{
-			name:        "file renamed",
-			expectError: true,
-			generateEvent: func(filePath string, cancel context.CancelFunc) {
-				os.Rename(filePath, filePath+"1")
-
-				// wait for the event to be handled
-				time.Sleep(1 * time.Second)
-				os.Remove(filePath + "1")
-			},
-		},
-		{
-			name:        "file removed",
-			expectError: true,
-			generateEvent: func(filePath string, cancel context.CancelFunc) {
-				// allow watcher handle to start
-				time.Sleep(1 * time.Second)
-				os.Remove(filePath)
-			},
+			name:     "with valid config",
+			filePath: "testdata/ec_config.yaml",
 		},
 	}
 
-	for _, testCase := range testCases {
-		t.Run(testCase.name, func(t *testing.T) {
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
 			ctx, cancel := context.WithCancel(context.Background())
-			t.Cleanup(cancel)
-			testFilePath := copyFileForTest(t, "testdata/ec_config.yaml")
+			defer cancel()
 
 			d := NewDynamicEncryptionConfiguration(
-				testCase.name,
-				testFilePath,
+				"test",
+				tt.filePath,
 				nil,
 				"",
 				"",
 			)
 
-			errs := make(chan error, 1)
 			go func() {
-				err := d.watchEncryptionConfigFile(ctx)
-				errs <- err
+				d.Run(ctx)
 			}()
 
-			testCase.generateEvent(d.filePath, cancel)
+			// allow the Run method to start
+			time.Sleep(time.Second)
 
-			err := <-errs
-			if testCase.expectError && err == nil {
-				t.Fatalf("expected error but got none")
-			}
-			if !testCase.expectError && err != nil {
-				t.Fatalf("expected no error but got %v", err)
+			// cancel the context to stop the Run method
+			cancel()
+
+			// allow the Run method to shut down
+			time.Sleep(time.Second)
+
+			// check if the queue has been shut down
+			if d.queue.ShuttingDown() != true {
+				t.Fatalf("expected queue to be shut down, but it was not")
```

Comment:
Avoid `sleep`, wrap the queue so that you can detect when `Done` has been called.